### PR TITLE
update: make mobile number optional during registeration

### PR DESCRIPTION
**Pull request by sumit4613**
# Hey, I just made a Pull Request!

## Description

Now the mobile number is not a required field for creating a user. I have updated the register API a little bit so that users can be registered without a mobile number.
## Motivation and Context

As @iamhssingh told that some user may don't want to share their mobile numbers that's why I made mobile optional.
## Have you tested this? If so, how?

I have written test case to verify that user can register without mobile number.
## Checklist for PR author(s)

- [x] Changes are covered by unit tests (no major decrease in code coverage %).
- [x] All tests pass.
- [x] Docstring coverage is **100%** via `interrogate drf_user` (I mean, we _should_ set a good example :smile:).
- [ ] Updates to documentation:
    - [ ] Document any relevant additions/changes in `README.md`.
    - [ ] Manually update **both** the `README.md` _and_ `docs/index.rst` for any new changes.
    - [ ] Any changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in the project's [``__init__.py``](https://github.com/101loop/drf-user/blob/master/drf_user/__init__.py) file.

## Release note

```release-note

```

### CODE REVIEW

File: drf_user/views.py [hunk 1]
```diff
@@ -50,14 +50,17 @@ class RegisterView(CreateAPIView):
...
+            "username": serializer.validated_data["username"],
+            "email": serializer.validated_data["email"],
+            "name": serializer.validated_data["name"],
+            "password": serializer.validated_data["password"],
+        }
+        try:
+            data["mobile"] = serializer.validated_data["mobile"]
+            return User.objects.create_user(**data)
+        except KeyError:
+            return User.objects.create_user(**data)
```
Comment (iamhssingh) [inline drf_user/views.py:63]:
```suggestion
        try:
            data["mobile"] = serializer.validated_data["mobile"]
        except KeyError:
            pass
        return User.objects.create_user(**data)
```

**Review by iamhssingh — COMMENTED** [review]
Shouldn't we have a Django config to make mobile compulsory, if required by any library user?

File: drf_user/managers.py [hunk 2]
```diff
@@ -15,7 +15,15 @@ class UserManager(BaseUserManager):
...
 
-    def _create_user(self, username, email, password, fullname, mobile, **extra_fields):
+    def _create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        fullname: str,
+        mobile: str = None,
+        **extra_fields: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:25]:
```suggestion
        **kwargs
```

File: drf_user/managers.py [hunk 3]
```diff
@@ -31,23 +39,31 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
 
-    def create_user(self, username, email, password, name, mobile, **extra_fields):
+    def create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
+        **extra_fields: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:49]:
```suggestion
        **kwargs
```

File: drf_user/managers.py [hunk 4]
```diff
@@ -59,7 +75,15 @@ def create_user(self, username, email, password, name, mobile, **extra_fields):
...
 
-    def create_superuser(self, username, email, password, name, mobile, **extra_fields):
+    def create_superuser(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
+        **extra_fields: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:85]:
```suggestion
        **kwargs
```

File: drf_user/views.py [hunk 5]
```diff
@@ -50,14 +52,23 @@ class RegisterView(CreateAPIView):
...
+        if (
+            not settings.USER_SETTINGS["MOBILE_OPTIONAL"]
+            and "mobile" not in serializer.validated_data
+        ):
+            raise ValidationError({"error": "Mobile is required."})
+
+        try:
+            data["mobile"] = serializer.validated_data["mobile"]
+        except KeyError:
+            pass
```
Comment (iamhssingh) [inline drf_user/views.py:70]:
```suggestion
        try:
            data["mobile"] = serializer.validated_data["mobile"]
        except KeyError:
            if not settings.USER_SETTINGS["MOBILE_OPTIONAL"]:
                raise ValidationError({"error": "Mobile is required."})
```

**Review by iamhssingh — CHANGES_REQUESTED** [review]

File: drf_user/managers.py [hunk 6]
```diff
@@ -15,7 +15,15 @@ class UserManager(BaseUserManager):
...
 
-    def _create_user(self, username, email, password, fullname, mobile, **extra_fields):
+    def _create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        fullname: str,
+        mobile: str = None,
+        **kwargs: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:25]:
```suggestion
        **kwargs: Dict
```
`dict` is not used in the type annotation

Comment (iamhssingh) [reply drf_user/managers.py:25]: Also, consider removing it as `**kwargs` is self-explanatory

File: drf_user/managers.py [hunk 7]
```diff
@@ -25,41 +33,55 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
 
-    def create_user(self, username, email, password, name, mobile, **extra_fields):
+    def create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
+        **kwargs: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:49]: Same as above ^

File: drf_user/managers.py [hunk 8]
```diff
@@ -25,41 +33,55 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
         Parameters
         ----------
         username: str
         email: str
         password: str
         name: str
-        mobile: str
-        extra_fields: dict
+        mobile: str, optional
+        kwargs: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:62]: Same as above ^

File: drf_user/managers.py [hunk 9]
```diff
@@ -25,41 +33,55 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
+        return self._create_user(username, email, password, name, mobile, **kwargs)
+
+    def create_superuser(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
+        **kwargs: dict
```
Comment (iamhssingh) [inline drf_user/managers.py:83]: Same as above

File: drf_user/managers.py [hunk 10]
```diff
@@ -25,41 +33,55 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
+
+        return self._create_user(username, email, password, name, mobile, **kwargs)
+
+    def create_superuser(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
```
Comment (iamhssingh) [inline drf_user/managers.py:82]:
```suggestion
        mobile: Optional[str] = None,
```

File: drf_user/managers.py [hunk 11]
```diff
@@ -25,41 +33,55 @@ def _create_user(self, username, email, password, fullname, mobile, **extra_fiel
...
         return user
 
-    def create_user(self, username, email, password, name, mobile, **extra_fields):
+    def create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        name: str,
+        mobile: str = None,
```
Comment (iamhssingh) [inline drf_user/managers.py:48]:
```suggestion
        mobile: Optional[str] = None,
```

File: drf_user/managers.py [hunk 12]
```diff
@@ -15,7 +15,15 @@ class UserManager(BaseUserManager):
...
     use_in_migrations = True
 
-    def _create_user(self, username, email, password, fullname, mobile, **extra_fields):
+    def _create_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        fullname: str,
+        mobile: str = None,
```
Comment (iamhssingh) [inline drf_user/managers.py:24]:
```suggestion
        mobile: Optional[str] = None,
```

File: drf_user/managers.py [hunk 13]
```diff
@@ -70,26 +92,22 @@ def create_superuser(self, username, email, password, name, mobile, **extra_fiel
...
-        Author: Himanshu Shankar (https://himanshus.com)
         """
         vals = update_user_settings()
 
-        extra_fields.setdefault("is_superuser", True)
-        extra_fields.setdefault("is_staff", True)
-        extra_fields.setdefault("is_active", vals.get("DEFAULT_ACTIVE_STATE", False))
+        kwargs.setdefault("is_superuser", True)
+        kwargs.setdefault("is_staff", True)
+        kwargs.setdefault("is_active", vals.get("DEFAULT_ACTIVE_STATE", False))
```
Comment (iamhssingh) [inline drf_user/managers.py:105]: What does the Python convention talk about modifying kwargs?

**Review by iamhssingh — CHANGES_REQUESTED** [review]
minor mistakes

File: drf_user/managers.py (hunk 13 above)
Comment (sumit4613) [reply drf_user/managers.py:107]: https://stackoverflow.com/questions/45883655/is-it-always-safe-to-modify-the-kwargs-dictionary

**Review by iamhssingh — APPROVED** [review]

**Pull request merged**